sample-nginx.conf

```diff
@@ -34,29 +34,39 @@ http {
         listen          80 default_server;
         server_name     _;
 
-        rewrite ^       https://stakepool.domain.tld/$request_uri permanent;
+        return 308      https://$host$request_uri;
     }
 
     limit_req_zone $binary_remote_addr zone=stakepool:10m rate=1r/s;
 
     server {
-        listen       443 default_server;
+        listen       443 http2 ssl default_server;
         server_name  _;
 
-        ssl                     on;
+        gzip on;
+        gzip_proxied any;
+        gzip_comp_level 3;
+        gzip_min_length 512;
+                gzip_types text/css text/* text/javascript application/x-javascript application/json
+                application/xml application/atom+xml application/xaml+xml application/javascript
+                application/x-font-ttf application/font-woff font/collection font/opentype
+                font/otf font/ttf image/bmp image/svg+xml image/x-icon application/octet-stream;
+
+        #ssl                     on; # only needed for older nginx where not in listen
         ssl_certificate         /etc/ssl/www/stakepool.domain.tld.crt;
         ssl_certificate_key     /etc/ssl/www/stakepool.domain.tld.key;
  
         ssl_session_cache               shared:SSL:20m;
-        ssl_protocols                   TLSv1 TLSv1.1 TLSv1.2;
+        ssl_protocols                   TLSv1.1 TLSv1.2; # add TLSv1.3 if supported
         ssl_ciphers                     EECDH+CHACHA20:EECDH+CHACHA20-draft:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:!MD5;
         ssl_prefer_server_ciphers       on;
+        #ssl_dhparam                     /etc/nginx/dhparam.pem; # openssl dhparam -out /etc/nginx/dhparam.pem 4096
         add_header                      Strict-Transport-Security max-age=15552001;
 
         # Serve static resources directly.
-        location /assets/* {
+        location /assets/ {

```

Comment:
What does the nginx error log say is happening here?  Are you sure this isn't the root vs. alias creating paths with "assets" in them.

But nginx paths do not need to end in a wild card like that.  Unless you are defining a regexp, which this does not, the star should do nothing.

Reply:
<details closed>
<summary>Error log</summary>
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/css/styles.css" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/css/styles.css HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/2-panel-mobile.png" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/2-panel-mobile.png HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/1-panel-mobile.png" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/1-panel-mobile.png HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/jquery.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/jquery.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/popper.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/popper.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/flickity.pkgd.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/flickity.pkgd.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/bootstrap.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/bootstrap.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/jquery.dataTables.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/jquery.dataTables.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/dataTables.bootstrap4.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/dataTables.bootstrap4.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/d3.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/d3pie.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3pie.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/global.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/global.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/3-panel-mobile.png" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/3-panel-mobile.png HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/group-1079.svg" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/group-1079.svg HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/circular-graphic-with-quarter-portion.svg" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/circular-graphic-with-quarter-portion.svg HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/servers.svg" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/servers.svg HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/status.svg" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/status.svg HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/flickity.pkgd.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/flickity.pkgd.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/jquery.dataTables.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/jquery.dataTables.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/dataTables.bootstrap4.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/dataTables.bootstrap4.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:34 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/d3.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:35 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/vendor/d3pie.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3pie.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:35 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/js/global.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/global.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:35 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/favicon/favicon-16x16.png" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/favicon/favicon-16x16.png?v=yHh3NA HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:35 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/favicon/favicon-32x32.png" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/favicon/favicon-32x32.png?v=yHh3NA HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 22:48:35 [error] 15637#15637: *3 open() "/opt/dcrstakepool/public/images/favicon/favicon.ico" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/images/favicon/favicon.ico?v=yHh3NA HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 23:48:20 [error] 15887#15887: *1 open() "/opt/dcrstakepool/public/css/styles.css" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/css/styles.css HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 23:48:20 [error] 15887#15887: *1 open() "/opt/dcrstakepool/public/js/vendor/d3.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
2019/09/24 23:48:20 [error] 15887#15887: *1 open() "/opt/dcrstakepool/public/js/vendor/d3pie.min.js" failed (2: No such file or directory), client: 153.248.16.27, server: _, request: "GET /assets/js/vendor/d3pie.min.js HTTP/2.0", host: "www.dcrstakedinner.com", referrer: "https://www.dcrstakedinner.com/"
</details>

Reply:
`open() "/opt/dcrstakepool/public/images/2-panel-mobile.png" failed (2: No such file or directory)`

Isn't that the correct path?  Or does your deployment actually have the assets path on disk?

Reply:
No.  I've always left the assets in the dcrstakepool directory.  Is the correct procedure to copy assets to /opt?  I don't believe I've seen those instructions anywhere.

Reply:
If I make it a relative path `./assets` it works.

Reply:
That's more than likely because it breaks the location entirely and the assets get served via dcrstakepool.  There are no wildcards (unless using the regexp syntax) or relative paths in an nginx location: https://nginx.org/en/docs/http/ngx_http_core_module.html#location

Reply:
> No. I've always left the assets in the dcrstakepool directory. Is the correct procedure to copy assets to /opt? I don't believe I've seen those instructions anywhere.

The public and views folders should be with the dcrstakepool binary.  But here we're talking about nginx and where it looks for it.  So I'm just confused where on disk you have public.  Do you mean you kept them in the github source folder?  The recommended way to run dcrstakepool is to copy (1) the binary, dcrstakepool, (2) the public folder, and (3) the views folder to some installation folder.  In the example config, that's `/opt/dcrstakepool`, but you have to make it and copy everything there.

Reply:
Updated README.md and sample-nginx.conf with more info on this.

Reply:
(It seems there's a lot more undocumented tribal knowledge than I thought.  Sorry bout that)

Reply:
After changing `alias /opt/dcrstakepool/public/;` to an absolute path pointing to that folder, and changing the permissions on that folder to include nginx's group, the errors disappeared.

Reply:
> Do you mean you kept them in the github source folder?

Yes.